tests/test_utility.py

```diff
@@ -20,6 +20,132 @@ def test_invalid_filetype_returns_informative_error(filename):
     assert "Invalid file type" in upload_error
 
 
+@pytest.mark.parametrize(
+    "original_df,duplicates_df",
+    [
+        (
+            pd.DataFrame(
+                {
+                    "participant_id": ["sub-1", "sub-1", "sub-2", "sub-2"],
+                    "session": [1, 2, 1, 2],
+                    "assessment_name": ["moca", "moca", "moca", "moca"],
+                    "assessment_score": [21.0, 24.0, np.nan, 24.0],
+                }
+            ),
+            # Have to specify column dtypes as well because the df will otherwise not evaluate as equal to an empty subset of above df
+            pd.DataFrame(
+                {
+                    "participant_id": pd.Series([], dtype="object"),
+                    "session": pd.Series([], dtype="int64"),
+                    "assessment_name": pd.Series([], dtype="object"),
+                    "assessment_score": pd.Series([], dtype="float64"),
+                }
+            ),
+        ),
+        (
+            pd.DataFrame(
+                {
+                    "participant_id": ["sub-1", "sub-1", "sub-2", "sub-2"],
+                    "session": [1, 1, 1, 2],
+                    "assessment_name": ["moca", "moca", "moca", "moca"],
+                    "assessment_score": [21.0, 24.0, np.nan, 24.0],
+                }
+            ),
+            pd.DataFrame(
+                {
+                    "participant_id": ["sub-1", "sub-1"],
+                    "session": [1, 1],
+                    "assessment_name": ["moca", "moca"],
+                    "assessment_score": [21.0, 24.0],
+                }
+            ),
+        ),
+        (
+            pd.DataFrame(
+                {
+                    "participant_id": ["sub-1", "sub-1", "sub-2", "sub-2"],
+                    "session": [np.nan, np.nan, 1, 2],
+                    "assessment_name": ["moca", "moca", "moca", "moca"],
+                    "assessment_score": [21.0, 24.0, np.nan, 24.0],
+                }
+            ),
```

Comment:
out of curiosity: if it was

```python
pd.DataFrame(
                {
                    "participant_id": ["sub-1", "sub-1", "sub-2", "sub-2"],
                    "session": [1, np.nan, 1, 2],
                    "assessment_name": ["moca", "moca", "moca", "moca"],
                    "assessment_score": [21.0, 24.0, np.nan, 24.0],
                }
            ),
```
You would not want to have this show up as a duplicate. I.e. missing values are just "values" in this context.

Reply:
yes, that's the current behaviour!

I ended up adding a test case for this as well 🙂